core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java

```diff
@@ -759,21 +874,102 @@ public Object parseTo(RangeFieldType fieldType, XContentParser parser, boolean c
         }
 
+        // No need to take into account Range#includeFrom or Range#includeTo, because from and to have already been
+        // rounded up via parseFrom and parseTo methods.
+        public abstract BytesRef encodeRanges(Set<Range> ranges) throws IOException;
+
+        public abstract NormalizedRange encodeRange(Object from, Object to);
+
         public final String name;
         private final NumberType numberType;
+
+        static BytesRef encodeLongRanges(Set<Range> ranges) throws IOException {
+            final byte[] encoded = new byte[5 + ((5 + 9) * 2) * ranges.size()];
+            ByteArrayDataOutput out = new ByteArrayDataOutput(encoded);
+            out.writeVInt(ranges.size());
+            for (Range range : ranges) {
+                byte[] encodedFrom = encode(((Number) range.from).longValue());
+                out.writeVInt(encodedFrom.length);
+                out.writeBytes(encodedFrom, encodedFrom.length);
+                byte[] encodedTo = encode(((Number) range.to).longValue());
+                out.writeVInt(encodedTo.length);
+                out.writeBytes(encodedTo, encodedTo.length);
+            }
+            return new BytesRef(encoded, 0, out.getPosition());
+        }
+
+        private static byte[] encode(long l) {
+            int sign = 1; // means positive
+            if (l < 0) {
+                l = -1 - l;
+                sign = 0;
+            }
+            return encode(l, sign);
+        }
+
+        private static byte[] encode(double d) {
+            long l;
+            int sign;
+            if (d < 0.0) {
+                l = Double.doubleToRawLongBits(-1 - d);
+                sign = 0;
+            } else {
+                l = Double.doubleToRawLongBits(d);
+                sign = 1; // means positive
+            }
+            return encode(l, sign);
+        }
+
+        private static byte[] encode(long l, int sign) {
+            assert l >= 0;
+            int bits = 64 - Long.numberOfLeadingZeros(l);
+
+            int numBytes = (bits + 7) / 8; // between 0 and 8
+            byte[] encoded = new byte[1 + numBytes];
+            // encode the sign first to make sure positive values compare greater than negative values
+            // and then the number of bytes, to make sure that large values compare greater than low values
+            if (sign > 0) {
+                encoded[0] = (byte) ((sign << 4) | numBytes);
+            } else {
+                encoded[0] = (byte) ((sign << 4) | (8 - numBytes));
+            }
+            for (int b = 0; b < numBytes; ++b) {
+                if (sign == 1) {
+                    encoded[encoded.length - 1 - b] = (byte) (l >>> (8 * b));
+                } else if (sign == 0) {
+                    encoded[encoded.length - 1 - b] = (byte) (0xFF - (l >>> (8 * b)));
+                } else {
+                    throw new AssertionError();
+                }
+            }
+            return encoded;
+        }
+
```

Comment:
I'm wondering that putting these helpers into a dedicated class would make things easier to read?